analysis/runtime/src/runtime/backend.rs

```diff
@@ -0,0 +1,94 @@
+use enum_dispatch::enum_dispatch;
+use fs_err::File;
+use std::env;
+use std::io::BufWriter;
+use std::sync::mpsc::Receiver;
+
+use bincode;
+
+use super::{AnyError, FINISHED};
+use crate::events::{Event, EventKind};
+use crate::metadata::{IWithMetadata, Metadata};
+
+#[enum_dispatch]
+trait IBackend {
+    fn write(&mut self, event: Event);
+}
+
+pub struct DebugBackend {
+    metadata: Metadata,
+}
+
+impl IBackend for DebugBackend {
+    fn write(&mut self, event: Event) {
+        eprintln!("{:?}", event.with_metadata(&self.metadata));
+    }
+}
+
+pub struct LogBackend {
+    writer: BufWriter<File>,
+}
+
+impl IBackend for LogBackend {
+    fn write(&mut self, event: Event) {
+        bincode::serialize_into(&mut self.writer, &event).unwrap();
+    }
+}
+
+#[enum_dispatch(IBackend)]
+pub enum Backend {
+    Debug(DebugBackend),
+    Log(LogBackend),
+}
+
+impl Backend {
+    fn write_all(&mut self, rx: Receiver<Event>) {
+        for event in rx {
+            if matches!(event.kind, EventKind::Done) {
+                return;
+            }
+            self.write(event);
+        }
+    }
+
+    pub fn run(&mut self, rx: Receiver<Event>) {
+        let (lock, cvar) = &*FINISHED;
+        let mut finished = lock.lock().unwrap();
+        self.write_all(rx);
+        *finished = true;
+        cvar.notify_one();
+    }
+}
+
+impl DebugBackend {
+    pub fn detect() -> Result<Self, AnyError> {
```

Comment:
why isn't this part of the `IBackend` interface?

Reply:
I originally tried to put it in `IBackend`, too, but that didn't work.